src/sparta/checkers/SimpleFlowAnnotatedTypeFactory.java

```diff
@@ -335,13 +327,55 @@ public AnnotationMirror getTopAnnotation(AnnotationMirror start) {
 
         @Override
         public AnnotationMirror getBottomAnnotation(AnnotationMirror start) {
-            if (start.toString().contains("Sink")) {
+            if (IFlowUtils.isSink(start)) {
                 return ANYSINK;
             } else {
                 return NOSOURCE;
             }
         }
 
+        @Override
+        public @Nullable AnnotationMirror leastUpperBound(AnnotationMirror a1, AnnotationMirror a2) {
+            if (!AnnotationUtils.areSameByName(getTopAnnotation(a1), getTopAnnotation(a2))) {
+                return null;
+            } else if (isSubtype(a1, a2)) {
+                return a2;
+            } else if (isSubtype(a2, a1)) {
+                return a1;
+            } else if (isSourceQualifier(a1) && isSourceQualifier(a2)) {
```

Comment:
Isn't checking both a1 and a2 redundant, because the first condition checked that both are from the same hierarchy? Otherwise the assert in the else branch wouldn't cover the case that a1 is a source and a2 is a sink, right?

Reply:
That's right! Thanks for pointing it out!